### ISSUE TITLE
[FSDP] - collective operation timeout (AllGather,Reduce) results in 5/5 training runs crashing at end of training loop (epoch 6+)

### ISSUE THREAD

**Issue by lessw2020** [pytorch/pytorch#77562]
### 🐛 Describe the bug

Running training on 2 different machines for various experiments with T5, 3B and 150K dataset. EC2, A100, PyTorch Nightly 514.
Training crashes with hang after completing training step for epoch 6 or 7. 
Of note - if you look at the saving info, you'll see that rank 0 fails to complete it's save in all these cases before the timeout is hit, meaning that may be what is causing the timeout (?).

<img width="1328" alt="another_hang" src="https://user-images.githubusercontent.com/46302957/168639609-5f034c9b-73e2-4bc0-a7f3-88ecde042939.png">
<img width="1332" alt="4of5_hang_end_of_training" src="https://user-images.githubusercontent.com/46302957/168639620-7c273a06-236a-4b4c-8da2-9c8d544f4ea8.png">
<img width="1449" alt="5of5_hang" src="https://user-images.githubusercontent.com/46302957/168639625-b340fd1f-4541-4a31-be5c-76bbc8fb7484.png">

At the end of training (training is done for that epoch, moving into validation loop) after epoch 6+, receive notice about :

[E ProcessGroupNCCL.cpp:737] [Rank 3] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=3738254, OpType=_ALLGATHER_BASE, Timeout(ms)=1800000) ran for 1809440 milliseconds before timing out.
[E ProcessGroupNCCL.cpp:737] [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=3738254, OpType=REDUCE, Timeout(ms)=1800000) ran for 1809061 milliseconds before timing out.

which then results in the training being shut down with
~~~
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 77665 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 77666 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 77667 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 77668 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 77669 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 77670 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 77671 closing signal SIGTERM

ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: -6) local_rank: 0 (pid: 77664) of binary: /home/ubuntu/anaconda3/envs/pytorch_p38/bin/python
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/envs/pytorch_p38/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 345, in wrapper
    return f(*args, **kwargs)
  File "/home/ubuntu/anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/torch/distributed/run.py", line 761, in main
    run(args)
  File "/home/ubuntu/anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/torch/distributed/run.py", line 752, in run
    elastic_launch(
  File "/home/ubuntu/anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/torch/distributed/launcher/api.py", line 131, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/ubuntu/anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/torch/distributed/launcher/api.py", line 245, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
main_childtuning.py FAILED
~~~

Note that I ran variations of this training 5 times, 2 different machines.  Every one crashed out at end of epoch 6 or 7, (was supposed to go to epoch 14). 
I encountered this with and without the dist.barrier() before the validation step. 

I've attached a pdf of one of the full crash stacks:
[training hang epoch 8.pdf](https://github.com/pytorch/pytorch/files/8701820/training.hang.epoch.8.pdf)

### Versions

PyTorch version: 1.12.0.dev20220514+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.22.4
Libc version: glibc-2.27

Python version: 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:59:51)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-1072-aws-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] numpydoc==1.1.0
[pip3] torch==1.12.0.dev20220514+cu113
[pip3] torch-model-archiver==0.5.0b20211117
[pip3] torch-workflow-archiver==0.2.0b20211118
[pip3] torchserve==0.5.0b20211117
[pip3] torchtext==0.11.0
[pip3] torchvision==0.13.0.dev20220514+cu113
[pip3] vit-pytorch==0.35.2
[conda] blas                      1.0                         mkl  
[conda] captum                    0.4.1                         0    pytorch
[conda] cudatoolkit               11.1.1               h6406543_9    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            12_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            12_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            12_linux64_mkl    conda-forge
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h95df7f1_0    conda-forge
[conda] mkl_fft                   1.3.1            py38h8666266_1    conda-forge
[conda] mkl_random                1.2.2            py38h1abd341_0    conda-forge
[conda] mypy_extensions           0.4.3            py38h578d9bd_4    conda-forge
[conda] numpy                     1.21.2           py38h20f2e39_0  
[conda] numpy-base                1.21.2           py38h79a1101_0  
[conda] numpydoc                  1.1.0                      py_1    conda-forge
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.12.0.dev20220514+cu113          pypi_0    pypi
[conda] torch-model-archiver      0.5.0                    py38_0    pytorch
[conda] torch-workflow-archiver   0.2.0                    py38_0    pytorch
[conda] torchserve                0.5.0                    py38_0    pytorch
[conda] torchtext                 0.11.0                     py38    pytorch
[conda] torchvision               0.13.0.dev20220514+cu113          pypi_0    pypi
[conda] vit-pytorch               0.35.2                   pypi_0    pypi

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by lessw2020** [comment]
update - verified 2x that running without any model checkpointing (currently saves the models at the end of training if it hits a new validation high) completes with no issue.
Noted that in all cases where it hangs, rank 0 never completes it's state dict, perhaps since it's streaming all of them to the CPU. 
![Uploading new_record_acc.png…]()

**Comment by pritamdamania87** [comment]
@lessw2020 One additional tool to debug this would be to use `TORCH_DISTRIBUTED_DEBUG=DETAIL`

cc @rohan-varma @zhaojuanmao regarding FSDP

**Comment by lessw2020** [comment]
> @lessw2020 One additional tool to debug this would be to use `TORCH_DISTRIBUTED_DEBUG=DETAIL`
> 
> cc @rohan-varma @zhaojuanmao regarding FSDP

Awesome, will re-run with that enabled. Thanks @pritamdamania87!

**Comment by rohan-varma** [comment]
@pritamdamania87 TORCH_DISTRIBUTED_DEBUG=DETAIL might not be too helpful yet as we use _allgather_base and _reduce_scatter_base which the ProcessGroupWrapper currently does not override. We should add these for better debugabilty.

**Comment by lessw2020** [comment]
I've been working with @rohan-varma  on this and have isolated things down to where can now reproduce in < 1 hour using the small dataset.
The core conflict revolves around a situation where the save at the end of an epoch is only happening based on a new high (low) being made in validation loss.
For scenarios where it is evaluation the 'current loss vs best_current_loss' if it does not save, then that is where the hang can occur.
If you run a training run with every epoch saving no matter what, then no hang.
A working theory would be that there is a dist.reduce call being made to finalize the 'current loss' and if ranks are out of synch on the resolution of that, then 1 or more ranks may drop into the 'time to save' mode, while others continue into the next epoch training - and hence the hang.
Will move the current debug dist.barrier() above this eval decision and see if that avoids the hang to try and refine this theory.
hang info is available with this internal paste: 
P502953327

**Comment by lessw2020** [comment]
Solved!!
@rohan-varma worked with me and ultimately he spotted the core issue here: 
the training code was using a dist.reduce for the ReduceOp.SUM to get the loss, but we needed to have a dist.*all*_reduce(op = ReduceOp.SUM). 
<img width="656" alt="training_hang_fix" src="https://user-images.githubusercontent.com/46302957/169908955-6120a17f-cf18-4628-8cb4-dc4620b76e08.png">

Basically my training code was only saving when the latest epoch val loss was a new low...and so the hang would occur between epochs 4-9, randomly, as only when this not a new low condition would not be met (current epoch did not set a new low).
The previous code reduce was only relevant for rank 0...so other ranks might have a different result for the loss and move into the save state, with the results of some moving to training and others moving to save...and thus the hang.

This also explains why no one else was hitting this issue - if you are a)never saving or b)always saving, then you have no branching in your code such that the rank disorder is exposed. 

Note that the FSDP tutorial needs to be updated for this fix b/c it has this same single reduce rather than the needed all_reduce, but will track that separately.

For now closing this issue with the fix being noted for anyone writing training code with FSDP.
